simple_parsing/helpers/partial.py

```diff
@@ -0,0 +1,308 @@
+from __future__ import annotations
+
+import dataclasses
+import functools
+import inspect
+import typing
+from dataclasses import make_dataclass
+from functools import lru_cache, singledispatch, wraps
+from logging import getLogger as get_logger
+from typing import (
+    Any,
+    Callable,
+    Generic,
+    Hashable,
+    Sequence,
+    TypeVar,
+    _ProtocolMeta,
+    cast,
+    get_type_hints,
+    overload,
+)
+
+from typing_extensions import ParamSpec
+
+import simple_parsing
+
+__all__ = ["Partial", "config_dataclass_for", "infer_type_annotation_from_default"]
+
+C = TypeVar("C", bound=Callable)
+_P = ParamSpec("_P")
+_T = TypeVar("_T", bound="Any")
+_C = TypeVar("_C", bound=Callable[..., Any])
+
+logger = get_logger(__name__)
+
+
+@singledispatch
+def infer_type_annotation_from_default(default: Any) -> Any | type:
+    if isinstance(default, (int, str, float, bool)):
+        return type(default)
+    if isinstance(default, tuple):
+        return typing.Tuple[tuple(infer_type_annotation_from_default(d) for d in default)]
+    if isinstance(default, list):
+        if not default:
+            return list
+        # Assuming that all items have the same type.
+        return typing.List[infer_type_annotation_from_default(default[0])]
+    if isinstance(default, dict):
+        if not default:
+            return dict
+    raise NotImplementedError(
+        f"Don't know how to infer type annotation to use for default of {default}"
+    )
+
+
+@singledispatch
+def adjust_default(default: Any) -> Any:
+    """Used the adjust the default value of a parameter that we extract from the signature.
+
+    IF in some libraries, the signature has a special default value, that we shouldn't use as the
+    default, e.g. "MyLibrary.REQUIRED" or something, then a handler can be registered here to
+    convert it to something else.
+    """
+    return default
+
+
+@overload
+def cache_when_possible(fn: _C) -> _C:
+    ...
+
+
+@overload
+def cache_when_possible(*, cache_fn=lru_cache) -> Callable[[_C], _C]:
+    ...
+
+
+@overload
+def cache_when_possible(fn: _C, *, cache_fn: Callable = lru_cache) -> _C:
+    ...
+
+
+def _default_cache_fn(fn: _C) -> _C:
+    return lru_cache(maxsize=None)(fn)  # type: ignore
+
+
+def cache_when_possible(
+    fn: _C | None = None, *, cache_fn: Callable = _default_cache_fn
+) -> _C | Callable[[_C], _C]:
+
+    if fn is None:
+
+        def _wrapper(_fn: _C) -> _C:
+            return cache_when_possible(_fn, cache_fn=cache_fn)
+
+        return _wrapper
+
+    cached_fn = cache_fn(fn)
+
+    @wraps(fn)
+    def _switch(*args, **kwargs):
+        if all(isinstance(arg, Hashable) for arg in args) and all(
+            isinstance(arg, Hashable) for arg in kwargs.values()
+        ):
+            return cached_fn(*args, **kwargs)
+        return fn(*args, **kwargs)
+
+    return _switch
+
+
+@cache_when_possible()
+def config_dataclass_for(
```

Comment:
Do you think this is a little verbose? `sp.config` might be a little too opaque, but I'd lean towards `sp.config`.

Reply:
Personally I'd go for `config_for`, it's short and clear and doesn't get in the way of discoverability (putting in the word dataclass means it will come up when I search for dataclass in my IDE, for no reason)

Reply:
Done